hoomd/data/data_structures.py

```diff
@@ -0,0 +1,497 @@
+                self._list.append(_to_hoomd_data_structure(val, type_def, self))
+
+    def __getitem__(self, index):
+        """Get the value(s) associated with entry index of the list.
+
+        Supports slicing.
+        """
+        return self._list[index]
+
+    def __setitem__(self, index, value):
+        """Set an item(s) at index to the given value.
+
+        Support assignment with slices.
+        """
+        if isinstance(index, slice):
+            with self._buffer():
+                for i, v in zip(self._slice_to_iterable(index), value):
+                    self[i] = v
+        else:
+            type_def = self._get_type_def(index)
+            try:
+                validated_value = type_def(value)
+            except TypeConversionError as err:
+                raise TypeConversionError(
+                    "Error in setting item {} in list.".format(index)) from err
+            else:
+                # If we override a list item that is itself a
+                # _HOOMDDataStructures we disconnect the old structure to
+                # prevent updates from it to force sync the list.
+                if isinstance(self._list[i], _HOOMDDataStructures):
+                    self._list[i]._parent = None
+
+                self._list[i] = _to_hoomd_data_structure(
+                    validated_value, type_def, self)
+        self._update()
+
+    def _slice_to_iterable(self, slice_):
+        start, end, step = slice_.start, slice_.end, slice_.step
+        if start is None:
+            if step is not None and step < 0:
+                iterable_start = len(self) - 1
+            else:
+                iterable_start = 0
+        if end is None:
+            if step is not None and step < 0:
+                iterable_end = -1
+            else:
+                iterable_end = len(self)
+        step = 1 if step is None else step
+        return range(iterable_start, iterable_end, step)
+
+    def _get_type_def(self, index):
+        entry_index = index % len(self._type_definition)
+        return self._type_definition[entry_index]
+
+    def __delitem__(self, index):
+        """Delele the index-th item from the list.
+
+        Supports slice deletion.
+        """
+        val = self._list[index]
+        # Disconnect item from list
+        if isinstance(val, _HOOMDDataStructures):
+            val._parent = None
+        del self._list[index]
+        # This is required for list type definitions which have index dependent
+        # validation.
+        if len(self._type_definition) > 1:
+            try:
+                _ = self._type_definition(self._list)
+            except TypeConversionError as err:
+                raise RuntimeError("Deleting items from list has caused an "
+                                   "invalid state.") from err
+        self._update()
+
+    def __len__(self):
+        """Get the length of the list."""
+        return len(self._list)
+
+    def insert(self, index, value):
```

Comment:
Strong recommendation: match the Python docs for all your standard container docstrings. Matching established conventions is better than inventing your own (such as "index-th"). Reference: https://docs.python.org/3/library/stdtypes.html#mutable-sequence-types